File: src/main/java/mediathek/daten/DatenProg.java
package mediathek.daten;

import mediathek.tool.GuiFunktionenProgramme;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

import static mediathek.controller.starter.RuntimeExec.TRENNER_PROG_ARRAY;

public class DatenProg implements Comparable<DatenProg> {

    public static final int PROGRAMM_NAME = 0;
    public static final int PROGRAMM_ZIEL_DATEINAME = 1;
    public static final int PROGRAMM_PROGRAMMPFAD = 2;
    public static final int PROGRAMM_SCHALTER = 3;
    public static final int PROGRAMM_PRAEFIX = 4;
    public static final int PROGRAMM_SUFFIX = 5;
    public static final int PROGRAMM_RESTART = 6;
    public static final int PROGRAMM_DOWNLOADMANAGER = 7;

    public static final int MAX_ELEM = 8;
    public static final String TAG = "Programm";
    public static final String[] COLUMN_NAMES = {"Beschreibung", "Zieldateiname", "Programm",
        "Schalter", "Präfix", "Suffix", "Restart", "Downloadmanager"};
    public static final String[] XML_NAMES = {"Programmname", "Zieldateiname", "Programmpfad",
        "Programmschalter", "Praefix", "Suffix", "Restart", "Downloadmanager"};

    public String[] arr;

    public DatenProg() {
        initialize();
        arr[PROGRAMM_RESTART] = Boolean.toString(false);
        arr[PROGRAMM_DOWNLOADMANAGER] = Boolean.toString(false);
    }

    public DatenProg(String name, String programmpfad, String schalter, String restart, String downloadmanager) {
        initialize();
        arr[PROGRAMM_NAME] = name;
        arr[PROGRAMM_PROGRAMMPFAD] = programmpfad;
        arr[PROGRAMM_SCHALTER] = schalter;
        arr[PROGRAMM_RESTART] = restart.equals("") ? Boolean.toString(false) : restart;
        arr[PROGRAMM_DOWNLOADMANAGER] = downloadmanager.equals("") ? Boolean.toString(false) : downloadmanager;
    }

    public DatenProg copy() {
        DatenProg ret = new DatenProg();
        System.arraycopy(this.arr, 0, ret.arr, 0, arr.length);
        return ret;
    }

    public boolean isRestart() {
        if (arr[PROGRAMM_RESTART].equals("")) {
            return false;
        }
        return Boolean.parseBoolean(arr[PROGRAMM_RESTART]);
    }

    public boolean isDownloadManager() {
        if (arr[PROGRAMM_DOWNLOADMANAGER].equals("")) {
            return false;
        }
        return Boolean.parseBoolean(arr[PROGRAMM_DOWNLOADMANAGER]);
    }

    public boolean urlTesten(String url) {
        //prüfen ob das Programm zur Url passt
        boolean ret = false;
        if (url != null) {
            //Felder sind entweder leer oder passen
            if (GuiFunktionenProgramme.praefixTesten(this.arr[PROGRAMM_PRAEFIX], url, true)
                    && GuiFunktionenProgramme.praefixTesten(this.arr[PROGRAMM_SUFFIX], url, false)) {
                ret = true;
            }
        }
        return ret;
    }

    public String getProgrammAufruf() {
        return arr[PROGRAMM_PROGRAMMPFAD] + " " + arr[PROGRAMM_SCHALTER];
    }

    public String getProgrammAufrufArray() {
        String ret;
        ret = arr[DatenProg.PROGRAMM_PROGRAMMPFAD];
        String[] ar = arr[DatenProg.PROGRAMM_SCHALTER].split(" ");
        for (String s : ar) {
            ret = ret + TRENNER_PROG_ARRAY + s;
        }
        return ret;
    }

    public static String makeProgAufrufArray(String pArray) {
        String[] progArray = pArray.split(TRENNER_PROG_ARRAY);
        String execStr = "";
        for (String s : progArray) {
            execStr = execStr + s + " ";
        }
        execStr = execStr.trim(); // letztes Leerzeichen wieder entfernen
        return execStr;
    }

    @Override
    public String toString() {
        String ret = "";
        for (int i = 0; i < MAX_ELEM; ++i) {
            if (i == 0) {
                ret += "| ***|" + COLUMN_NAMES[i] + ": " + arr[i] + System.lineSeparator();
            } else {
                ret += "|    |" + COLUMN_NAMES[i] + ": " + arr[i] + System.lineSeparator();
            }
        }
        return ret;
    }

    private void initialize() {
        arr = new String[MAX_ELEM];
        Arrays.fill(arr,"");
    }

    @Override
    public int compareTo(@NotNull DatenProg o) {
        return 0;
    }
}

File: src/main/kotlin/mediathek/daten/DatenProg.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.daten

import mediathek.controller.starter.RuntimeExec
import mediathek.tool.GuiFunktionenProgramme

class DatenProg() {
    private var arr: Array<String> = defaultValues()

    constructor(
        name: String,
        programmpfad: String,
        schalter: String,
        restart: String,
        downloadmanager: String,
    ) : this() {
        this.name = name
        programPath = programmpfad
        switches = schalter
        arr[PROGRAMM_RESTART] = restart.ifEmpty { false.toString() }
        arr[PROGRAMM_DOWNLOADMANAGER] = downloadmanager.ifEmpty { false.toString() }
    }

    fun copy(): DatenProg {
        val ret = DatenProg()
        arr.copyInto(ret.arr)
        return ret
    }

    fun toArray(): Array<String> =
        arr.clone()

    fun copyFrom(values: Array<String>) {
        arr = defaultValues()
        values.copyInto(arr, endIndex = minOf(values.size, arr.size))
    }

    operator fun get(index: Int): String =
        arr[index]

    operator fun set(index: Int, value: String) {
        arr[index] = value
    }

    var name: String
        get() = arr[PROGRAMM_NAME]
        set(value) {
            arr[PROGRAMM_NAME] = value
        }

    var targetFileName: String
        get() = arr[PROGRAMM_ZIEL_DATEINAME]
        set(value) {
            arr[PROGRAMM_ZIEL_DATEINAME] = value
        }

    var programPath: String
        get() = arr[PROGRAMM_PROGRAMMPFAD]
        set(value) {
            arr[PROGRAMM_PROGRAMMPFAD] = value
        }

    var switches: String
        get() = arr[PROGRAMM_SCHALTER]
        set(value) {
            arr[PROGRAMM_SCHALTER] = value
        }

    var prefix: String
        get() = arr[PROGRAMM_PRAEFIX]
        set(value) {
            arr[PROGRAMM_PRAEFIX] = value
        }

    var suffix: String
        get() = arr[PROGRAMM_SUFFIX]
        set(value) {
            arr[PROGRAMM_SUFFIX] = value
        }

    var isRestart: Boolean
        get() = arr[PROGRAMM_RESTART].isNotEmpty() && arr[PROGRAMM_RESTART].toBoolean()
        set(value) {
            arr[PROGRAMM_RESTART] = value.toString()
        }

    var isDownloadManager: Boolean
        get() = arr[PROGRAMM_DOWNLOADMANAGER].isNotEmpty() && arr[PROGRAMM_DOWNLOADMANAGER].toBoolean()
        set(value) {
            arr[PROGRAMM_DOWNLOADMANAGER] = value.toString()
        }

    fun urlTesten(url: String?): Boolean =
        url != null &&
            GuiFunktionenProgramme.checkPrefix(prefix, url) &&
            GuiFunktionenProgramme.checkSuffix(suffix, url)

    val programmAufruf: String
        get() = "$programPath $switches"

    val programmAufrufArray: String
        get() = buildString {
            append(programPath)
            for (switch in switches.split(" ")) {
                append(RuntimeExec.TRENNER_PROG_ARRAY)
                append(switch)
            }
        }

    override fun toString(): String =
        buildString {
            for (i in 0 until MAX_ELEM) {
                if (i == 0) {
                    append("| ***|")
                } else {
                    append("|    |")
                }
                append(COLUMN_NAMES[i])
                append(": ")
                append(arr[i])
                append(System.lineSeparator())
            }
        }

    companion object {
        const val PROGRAMM_NAME = 0
        const val PROGRAMM_ZIEL_DATEINAME = 1
        const val PROGRAMM_PROGRAMMPFAD = 2
        const val PROGRAMM_SCHALTER = 3
        const val PROGRAMM_PRAEFIX = 4
        const val PROGRAMM_SUFFIX = 5
        const val PROGRAMM_RESTART = 6
        const val PROGRAMM_DOWNLOADMANAGER = 7

        const val MAX_ELEM = 8
        const val TAG = "Programm"

        val COLUMN_NAMES: Array<String> = arrayOf(
            "Beschreibung",
            "Zieldateiname",
            "Programm",
            "Schalter",
            "Präfix",
            "Suffix",
            "Restart",
            "Downloadmanager",
        )

        val XML_NAMES: Array<String> = arrayOf(
            "Programmname",
            "Zieldateiname",
            "Programmpfad",
            "Programmschalter",
            "Praefix",
            "Suffix",
            "Restart",
            "Downloadmanager",
        )

        fun makeProgAufrufArray(pArray: String): String =
            pArray.split(RuntimeExec.TRENNER_PROG_ARRAY)
                .joinToString(separator = " ")
                .trim()

        private fun defaultValues(): Array<String> =
            Array(MAX_ELEM) { "" }.apply {
                this[PROGRAMM_RESTART] = false.toString()
                this[PROGRAMM_DOWNLOADMANAGER] = false.toString()
            }
    }
}
